### PR TITLE
Add integration test to run a helm deploy pipeline

### PR DESCRIPTION
**Pull request by pivotal-nader-ziada**
Related to: #63 

## Proposed changes
 * Create a pipeline of two tasks,
 * First task build and push an image
  * Second task helm deploys that image
 * Verify the service created is reachable

/cc @bobcatfish

### CODE REVIEW

File: test/crd_checks.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 
 const (
 	interval = 1 * time.Second
-	timeout  = 2 * time.Minute
+	timeout  = 4 * time.Minute
```
Comment (bobcatfish) [inline test/crd_checks.go]: at this point, what do you think about adding a `timeout` argument to the polling functions, so that each test can pass its own timeout? e.g. the helloworld task probably needs < 1 minute, kaniko needs 2 minutes, this one needs 4 minutes. it will help to keep track of why the timeout is what it is, and also will ensure tests don't wait for longer than they need to

File: test/helm_task_test.go [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+		serviceIp = ingress[0].IP
+	}
+
+	resp, err := http.Get(fmt.Sprintf("http://%s:8080", serviceIp))
+	if err != nil {
+		t.Errorf("Error reaching service at http://%s:8080 %s", serviceIp, err)
+	}
+	if resp != nil && resp.StatusCode != http.StatusOK {
+		t.Errorf("Error from service at http://%s:8080 %s", serviceIp, err)
+	}
```
Comment (bobcatfish) [inline test/helm_task_test.go]: a beautiful test!!! :heart_eyes:

File: test/helm_task_test.go [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+					},
+				},
+			},
+			BuildSpec: &buildv1alpha1.BuildSpec{
+				Steps: []corev1.Container{{
+					Name:  "kaniko",
+					Image: "gcr.io/kaniko-project/executor",
+					Args: []string{"--dockerfile=/workspace/Dockerfile",
+						fmt.Sprintf("--destination=%s", imageName),
+					},
```
Comment (bobcatfish) [inline test/helm_task_test.go]: in the long run i feel like we can probably get rid of the kaniko specific integration test since we're using kaniko here too. @tanner-bruce is making some updates to it in #158 that we don't want to lose tho, so maybe we can do that in another PR later sometime.

File: test/helm_task_test.go [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+						Type: v1alpha1.PipelineResourceTypeGit,
+					},
+				},
+			},
+			BuildSpec: &buildv1alpha1.BuildSpec{
+				Steps: []corev1.Container{{
+					Name:  "kaniko",
+					Image: "gcr.io/kaniko-project/executor",
+					Args: []string{"--dockerfile=/workspace/Dockerfile",
+						fmt.Sprintf("--destination=%s", imageName),
```
Comment (bobcatfish) [inline test/helm_task_test.go]: if you wanted, we have templating working now, so you could use Task params for the image name and path

File: test/helm_task_test.go [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+						Name:  "helm-deploy",
+						Image: "alpine/helm",
+						Args: []string{"install",
+							"--debug",
+							"--name=${inputs.params.chartname}",
+							"${inputs.params.pathToHelmCharts}",
+							"--set",
+							"image.repository=${inputs.params.image}",
+						},
+					}},
```
Comment (bobcatfish) [inline test/helm_task_test.go]: i think something is off with the indentation b/w these

File: test/helm_task_test.go [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+				}, {
+					Name: "chartname",
+				}},
+			},
+			BuildSpec: &buildv1alpha1.BuildSpec{
+				Steps: []corev1.Container{{
+					Name:  "helm-init",
+					Image: "alpine/helm",
+					Args:  []string{"init"},
+				},
```
Comment (bobcatfish) [inline test/helm_task_test.go]: i wonder if the teardown step in the test should invoke a Task or something to remove `helm`, so that ppl who run this test aren't left with helm running in their clusters (or will the namespace take care of that?)

File: test/helm_task_test.go [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+							"--set",
+							"image.repository=${inputs.params.image}",
+						},
+					}},
+			},
+		},
+	}
+}
+
+func getelmDeployPipeline(namespace string) *v1alpha1.Pipeline {
```
Comment (bobcatfish) [inline test/helm_task_test.go]: typo?

File: test/helm_task_test.go [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+						fmt.Sprintf("--destination=%s", imageName),
+					},
+				}},
+			},
+		},
+	}
+}
+
+func getHelmDeployTask(namespace string) *v1alpha1.Task {
+	return &v1alpha1.Task{
```
Comment (tejal29) [inline test/helm_task_test.go]: This is something, we haven't figured out yet. You can address it now or later.
Shd we host all the Tasks Defn somewhere for others to use? 
And then the tests should read the task def from that file. 
The reason why i think we shd do so is
- Users can see all the tasks def at one place 
-  In future, if we decide to make them public-ally available, there are no two copies and tests always test the public ones.

File: test/helm_task_test.go (hunk 7 above)
Comment (tejal29) [inline test/helm_task_test.go]: fix spell error.

File: test/helm_task_test.go [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+						Name:  "image",
+						Value: imageName,
+					}},
+				},
+			},
+		},
+	}
+}
+
+func getelmDeployPipelineRun(namespace string) *v1alpha1.PipelineRun {
```
Comment (tejal29) [inline test/helm_task_test.go]: fix spell error

**Comment by bobcatfish** [comment]
image
 
wut

/meow space

File: test/crd_checks.go [hunk 10]
```diff
@@ -30,7 +30,7 @@ import (
 
 const (
 	interval = 1 * time.Second
-	timeout  = 2 * time.Minute
+	timeout  = 5 * time.Minute
```
[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply test/crd_checks.go]: the test is running a pipeline of two tasks including pushing an image and waiting for the service to get an external ip, so it was timing out

File: test/helm_task_test.go [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+		t.Fatalf("Failed to create Task `%s`: %s", createImageTaskName, err)
+	}
+
+	logger.Infof("Creating Task %s", helmDeployTaskName)
+	if _, err := c.TaskClient.Create(getHelmDeployTask(namespace)); err != nil {
+		t.Fatalf("Failed to create Task `%s`: %s", helmDeployTaskName, err)
+	}
+
+	logger.Infof("Creating Pipeline %s", helmDeployPipelineName)
+	if _, err := c.PipelineClient.Create(getelmDeployPipeline(namespace)); err != nil {
```
[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: fixed

File: test/helm_task_test.go [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+	}, "PipelineRunCompleted"); err != nil {
+		t.Errorf("Error waiting for PipelineRun %s to finish: %s", helmDeployPipelineRunName, err)
+	}
+
+	k8sService, err := c.KubeClient.Kube.CoreV1().Services(namespace).Get(helmDeployServiceName, metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("Error getting service at %s %s", helmDeployServiceName, err)
+	}
+	var serviceIp string
+	ingress := k8sService.Status.LoadBalancer.Ingress
```
[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: will add a check

File: test/helm_task_test.go [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sourceResourceName,
+			Namespace: namespace,
+		},
+		Spec: v1alpha1.PipelineResourceSpec{
+			Type: v1alpha1.PipelineResourceTypeGit,
+			Params: []v1alpha1.Param{
+				v1alpha1.Param{
+					Name:  "Url",
+					Value: "https://github.com/pivotal-nader-ziada/gohelloworld",
```
[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: test code for what?

File: test/helm_task_test.go [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+func getCreateImageTask(namespace string, t *testing.T) *v1alpha1.Task {
+	// according to knative/test-infra readme (https://github.com/knative/test-infra/blob/13055d769cc5e1756e605fcb3bcc1c25376699f1/scripts/README.md)
+	// the KO_DOCKER_REPO will be set with according to the porject where the cluster is created
+	// it is used here to dunamically get the docker registery to push the image to
+	dockerRepo := os.Getenv("KO_DOCKER_REPO")
+	if dockerRepo == "" {
+		t.Fatalf("KO_DOCKER_REPO env variable is required")
+	}
+
+	imageName = fmt.Sprintf("%s/%s", dockerRepo, sourceImageName)
```
[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: ok, will append with random name

File: test/helm_task_test.go (hunk 3 above)
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: yes, agreed

File: test/helm_task_test.go (hunk 8 above)
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: I thought about this and here are my reason for keeping these tasks in this test:
 * the generic file with all tasks will grow too big and be hard to read
  * these tasks should be specific to the test using them and having them in a common file can make changing tests harder
  * I expect the public task definitions to be in yaml and in a different repo all together

Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: you make a valid point, its open for discussion for sure

File: test/helm_task_test.go (hunk 6 above)
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: doesn't the cluster get deleted after the test run in Prow?

File: test/helm_task_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: you mean add this helm chart and hello world sample in this same repo here? the build and push image was giving me issues when the I tired that, could not find the Dockerfile. Let me know if you want to use a different repo

[the remark this answers is not in the document]
Comment (adrcunha) [reply test/helm_task_test.go]: This test app should not be in an external repository, as it restricts our ability to make changes and fix bugs.

File: test/helm_task_test.go (hunk 6 above)
Comment (bobcatfish) [reply test/helm_task_test.go]: > I was worried if if things fail running another task might not work for whatever reason and I wasn't sure how else to remove it

That's true, so a couple options:
- Try to do it anyway but don't actually fail anything if the teardown fails (just log it)
- Try to do teardown without using Pipeline/Task, e.g. call out to kubectl

> doesn't the cluster get deleted after the test run in Prow?

Yeah I think so! So we're fine for Prow, I'm more concerned about ppl running this on their own clusters.

File: test/helm_task_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (bobcatfish) [reply test/helm_task_test.go]: > This test app should not be in an external repository, as it restricts our ability to make changes and fix bugs.

:woman_shrugging: we're okay with it, we could always copy the code over here if it became a problem

[the remark this answers is not in the document]
Comment (bobcatfish) [reply test/helm_task_test.go]: @tejal29 do you mean add some tests to `https://github.com/pivotal-nader-ziada/gohelloworld` which we could also execute with the pipeline/tasks? that would be cool!

File: test/helm_task_test.go (hunk 8 above)
Comment (bobcatfish) [reply test/helm_task_test.go]: We can keep iterating on this but what I found when working on knative/serving conformance tests was that using yaml directly was hard in tests b/c sometimes we'd need a field to be dynamic (e.g. namespaces)

So what I'd like to do in the long run is have both integration tests like these that use resources defined in code which cover very specific cases we want to test, and examples in `examples` which are yaml and tested (see https://github.com/knative/build-pipeline/issues/108).

I think you're right @tejal29 , we'll probably remove some of these integration test cases in the long run (maybe even just one, with a complex pipeline).

File: test/helm_task_test.go (hunk 6 above)
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: will get the test green first then add the cleanup

File: test/helm_task_test.go (hunk 8 above)
Comment (tejal29) [reply test/helm_task_test.go]: sounds good to me. We shd figure out a plan on where and how to make tasks available. 
Something like [krew](https://github.com/GoogleContainerTools/krew)
And then a plan to promote well tested tasks from here to this repo/location.
Also make sure, we don't make any backward incompatible changes to the controller and break existing tasks.

File: test/helm_task_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tejal29) [reply test/helm_task_test.go]: @pivotal-nader-ziada Sorry for not making it clear before. Agree with @adrcunha and i feel test app should remain in the same repo. Its easier to iterate over. Also our tests do not depend on a repo which is not under our control. Not suggesting you will be malicious and add commits which break our tests :)

I was wondering if we could copy this and add at testapp/go-hello-world/<>

File: test/crd_checks.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tejal29) [reply test/crd_checks.go]: ah ok.

File: test/helm_task_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tejal29) [reply test/helm_task_test.go]: if it turns out to be a longer turnaround time, i do not mind copying them over in another PR.

**Comment by pivotal-nader-ziada** [comment]
Added cleanup task after the test is runs to remove tiller from cluster

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go [hunk 15]
```diff
@@ -54,8 +54,32 @@ func GetNextTask(prName string, state []*PipelineRunTaskRun, logger *zap.Sugared
...
+						case corev1.ConditionFalse:
+							return false
+						case corev1.ConditionUnknown:
+							return false
+						}
+					}
+				}
+			}
+		}
+	}
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go]: whoa nice! im curious, did you need to add this for the helm deployment to work or was this just a bonus?

One thing I am a bit confused about: let's say we have these tasks:

1. Task A
2. Task B - has a passed constraint on task A

Since Task A has to run before Task B, how can this function ever return False? This might be more relevant once we are constructing a DAG instead of executing Tasks in the order they are declared in, but even then, I think the `passedConstraints` would be used to construct that DAG only.

Long story short, I think if this function returns `False`, it means the `PipelineRun` is in an error state and will never complete, and we need to return an error and stop reconciling the `PipelineRun` at that point (maybe I'm missing something tho!)

File: test/helm_task_test.go [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+		if resp != nil && resp.StatusCode != http.StatusOK {
+			t.Errorf("Error from service at http://%s:8080 %s", serviceIp, err)
+		}
+
+	} else {
+		t.Errorf("Service IP is empty.")
+	}
+
+	// cleanup task to remove helm from cluster, will not fail the test if it fails, just log
+	removeHelmFromCluster(c, t, namespace, logger)
```
Comment (bobcatfish) [inline test/helm_task_test.go]: i think we should make sure this runs if the test is interrupted or fails, e.g. like we do with the `teardown` (and/or move it to the teardown):

```
	knativetest.CleanupOnInterrupt(func() { tearDown(logger, c.KubeClient, namespace) }, logger)
	defer tearDown(logger, c.KubeClient, namespace)
```

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go (hunk 15 above)
Comment (pivotal-nader-ziada) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go]: I thought the test was failing because both tasks were running at the same time, and since the deploy one needs the image from the first one, I added this functionality 
my understanding was that the `cantaskrun` will return false if the task is not ready, but will get called again in the following reconcile period, am I wrong?

File: test/helm_task_test.go (hunk 16 above)
Comment (pivotal-nader-ziada) [reply test/helm_task_test.go]: ok

**Comment by bobcatfish** [comment]
Just a quick note that we should leave #63 open after this b/c we still need to do some work to respect the cluster parameters (e.g. mount and create a kubeconfig that helm can use?). At the moment, we can only deploy to the cluster that the Pipeline CRD is already running in.

(Apologies if I added this comment already, I thought I did but I can't find it 😅 )

**Comment by pivotal-nader-ziada** [comment]
@bobcatfish should the different cluster be a different issue since this is not yet implemented in the controllers?

File: pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go (hunk 15 above)
Comment (tejal29) [reply pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go]: umm. may be some misunderstanding.  The `GetNextTask` is still check if the status of the previous task run is complete.
Ideally, we would not care if the prev task is completed or not. 
```
func GetNextTask(prName string, state []*PipelineRunTaskRun, logger *zap.SugaredLogger) *PipelineRunTaskRun {
 	for _, prtr := range state {
 		if prtr.TaskRun != nil {
 			c := prtr.TaskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
 			if c == nil {
 				logger.Infof("TaskRun %s doesn't have a condition so it is just starting and we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
 				return nil
 			}
 			switch c.Status {
 			case corev1.ConditionFalse:
 				logger.Infof("TaskRun %s has failed; we don't need to run PipelineRun %s", prtr.TaskRunName, prName)
 				return nil
 			case corev1.ConditionUnknown:
				logger.Infof("TaskRun %s is still running so we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
				return nil
			}
		} else if canTaskRun(prtr.PipelineTask) {
		} else if canTaskRun(prtr.PipelineTask, state) {
			logger.Infof("TaskRun %s should be started for PipelineRun %s", prtr.TaskRunName, prName)
			return prtr
		}
}
```

**Comment by shashwathi** [comment]
This PR also addresses : https://github.com/knative/build-pipeline/issues/148

**Comment by bobcatfish** [comment]
> @bobcatfish should the different cluster be a different issue since this is not yet implemented in the controllers?

sure that's fine!
